### PR TITLE
Adding high-level documentation on the build system

### PR DESCRIPTION
**Pull request by arschles**
Content with help from @martinmaly

### CODE REVIEW

File: docs/BUILD_SYSTEM.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+Several components are built from this repository (catalog controller,
+example brokers, ...). One or more artifacts are constructed for each component
+(e.g. docker images, binaries). Generally speaking there is a single
+Makefile per component.
+
+## Structure
+
+There is a top-level [Makefile](./Makefile) that drives high-level actions
+(i.e. installing repository-wide Go dependencies) and recursively builds other
+targets in each of the directories listed in the DIRS variable.
```
Comment (martinmaly) [inline docs/BUILD_SYSTEM.md]: consider formatting the DIRS variable as `DIRS`

File: docs/BUILD_SYSTEM.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+For example, typing make build at the top level will effectively run
```
Comment (martinmaly) [inline docs/BUILD_SYSTEM.md]: we could format make build as preformatted using backticks:

`make build`

File: docs/BUILD_SYSTEM.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+`make -C ${DIR}` build for each DIR in DIRS.
+
+The subdirectory Makefiles are designed to be used both by calling
+`make <target>` from top level (in which case `make <target>` in all
+subdirectories will be executed) but also as standalone Makefiles. This is
+to make it possible to run make in each individual subdirectory, and
```
Comment (martinmaly) [inline docs/BUILD_SYSTEM.md]: line 28 - format make as `make`

File: docs/BUILD_SYSTEM.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+operate on subset of the code base.
+
+
+## Subdirectory Makefiles
+
+Each subdirectory that contains code that must be built into an artifact has
+a Makefile in it. Generally, these Makefiles are simple because they specify
```
Comment (martinmaly) [inline docs/BUILD_SYSTEM.md]: `Makefile` (backticks?)

File: docs/BUILD_SYSTEM.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+a few variables, and the remainder of the Makefile code is included from
+`hack/Makefile.mk` and `hack/Common.mk`.
```
Comment (martinmaly) [inline docs/BUILD_SYSTEM.md]: these could be links (makefile.mk and common.mk)

File: docs/BUILD_SYSTEM.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+## Common Code
+
+Two shared makefiles provide common functionality to all subdirectory
+Makefiles: [`Makefile.mk`](./hack/Makefile.mk) and
+[`Common.mk`](./hack.Common.mk). See below for a description of each common
+file:
+
+- [`Makefile.mk`](./hack/Makefile.mk) - top-level variable definitions (e.g.
+  the Go binary name, Go version and GOPATH) and macros to automate common
```
Comment (martinmaly) [inline docs/BUILD_SYSTEM.md]: consider formatting GOPATH as `GOPATH`

File: docs/BUILD_SYSTEM.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  functionality. These macros include but are not limited to:
+    - Determining whether a docker push should be executed based on parameters
+      in the subdirectory's Makefile. If so, executing the push
+    - Determining whether a docker-container-based binary build should be
+      executed based on parameters in the subdirectory's `Makefile`. If so,
+      executing the container-based build in the subdirectory's `Makefile`
+    - Defines the `all` target, which loves to be the first target defined in
```
Comment (martinmaly) [inline docs/BUILD_SYSTEM.md]: (my bad). the 'which loves to be' may sound a little informal. Maybe:

Defines the `all` target, in order to make if the first target defined the `Makefile`.

(we could even leave it at that)

File: docs/BUILD_SYSTEM.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+      the `Makefile` and is therefore defined in
+      [`Makefile.mk`](./hack/Makefile.mk) which is included first
+- [`Common.mk`](./hack/Common.mk) - defines all common build targets (e.g.
```
Comment (martinmaly) [inline docs/BUILD_SYSTEM.md]: would it damage formatting to insert a newline before - [`Common.mk`](...) ?

File: docs/BUILD_SYSTEM.md [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+  `build`, `test`) except for `all`. The common build targets use the
+  variables defined by the subdirectory's `Makefile` to define the make
+  targets. For example, if the `BIN` variable is set in a subdirectory's
+  `Makefile`, the build target will run an appropriate go build command.
```
Comment (martinmaly) [inline docs/BUILD_SYSTEM.md]: format go build via backticks: `go build`

**Comment by martinmaly** [comment]
LGTM (`LGTM1` applied)

File: docs/BUILD_SYSTEM.md [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+Makefile per component.
+
+## Structure
+
+There is a top-level [Makefile](./Makefile) that drives high-level actions
+(i.e. installing repository-wide Go dependencies) and recursively builds other
+targets in each of the directories listed in the DIRS variable.
+
+For example, typing make build at the top level will effectively run
+`make -C ${DIR}` build for each DIR in DIRS.
```
Comment (duglin) [inline docs/BUILD_SYSTEM.md]: s/DIR in DIRs/component of the project/

File: docs/BUILD_SYSTEM.md [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+## Subdirectory Makefiles
+
+Each subdirectory that contains code that must be built into an artifact has
+a Makefile in it. Generally, these Makefiles are simple because they specify
+a few variables, and the remainder of the Makefile code is included from
+`hack/Makefile.mk` and `hack/Common.mk`.
+
+Below is a rough outline of one of these subdirectory `Makefile`s (eliding the
+copyright notice):
+
```
Comment (duglin) [inline docs/BUILD_SYSTEM.md]: I would either talk about the general flow of Makefile.mk and Common.mk w/o including snippets or leave it out all together. The problem with having snippets of code here is that its sure to get out of sync with the real Makefile and then it loses its value and could just confuse people even more.  Plus, I'm not sure I see the value of showing how we set a few env var & some "include"s w/o any real logic.

File: docs/BUILD_SYSTEM.md [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+Two shared makefiles provide common functionality to all subdirectory
+Makefiles: [`Makefile.mk`](./hack/Makefile.mk) and
+[`Common.mk`](./hack.Common.mk). See below for a description of each common
+file:
+
+- [`Makefile.mk`](./hack/Makefile.mk) - top-level variable definitions (e.g.
+  the Go binary name, Go version and GOPATH) and macros to automate common
+  functionality. These macros include but are not limited to:
+    - Determining whether a docker push should be executed based on parameters
```
Comment (duglin) [inline docs/BUILD_SYSTEM.md]: Should we describe how it determines these things?

File: docs/BUILD_SYSTEM.md (hunk 1 above)
Comment (pmorie) [inline docs/BUILD_SYSTEM.md]: Agree

File: docs/BUILD_SYSTEM.md (hunk 3 above)
Comment (pmorie) [inline docs/BUILD_SYSTEM.md]: I strongly want to do away with these package level Makefiles, but I am glad that they are documented for now, at least.

File: docs/BUILD_SYSTEM.md (hunk 12 above)
Comment (pmorie) [inline docs/BUILD_SYSTEM.md]: I don't want to get into that level of detail, unless we want to be forever rewriting things.  That level of detail should live in the code, IMO.

File: docs/BUILD_SYSTEM.md (hunk 7 above)
Comment (pmorie) [inline docs/BUILD_SYSTEM.md]: Yeah, I prefer to keep things less conversational and editorial in doc like this.

File: docs/BUILD_SYSTEM.md [hunk 13]
```diff
@@ -0,0 +1,78 @@
+# The Service-Catalog Build System
+
+This document aims to overview the build system in this repository.
```
Comment (krancour) [inline docs/BUILD_SYSTEM.md]: "Overview" is used like a verb here, which is awkward.

Consider "This document presents an overview of..."

**Comment by krancour** [comment]
I like the information presented in here and I won't register too many nits because it looks like this is all pretty well dissected already. My one significant piece of feedback is that the structure  / outline of the doc could be improved.

Some concrete suggestions:

The h2 currently titled "Composition" might be better renamed as "Structure," with a goal of describing the structure in general terms. i.e. There's a top level Makefile and also one per constituent component. To accomplish this, there are a few introductory sentences from the subsequent sections that could be moved up into this section.

The h2 _currently_ titled "Structure" could become an h3 and be retitled as "Top Level Makefile."

The h2 titled "Subdirectory Makefiles" could become an h3.

The h2 titled "Common Code" could become an h3.

I would recommend adding a new h2 at the bottom and have that section contain crib notes or a cheat cheat on what targets are available and what they do... better yet, describe a reasonable workflow for people who are hacking on one or more components in this repo.

File: docs/BUILD_SYSTEM.md (hunk 8 above)
Comment (arschles) [inline docs/BUILD_SYSTEM.md]: are you talking about inserting a newline before the entire bullet point (the `-` character)?

Comment (martinmaly) [inline docs/BUILD_SYSTEM.md]: yes.

I know it sounds OCD :( personally I tend to view *.md files as though they should be readable as plain text as well as rendered by something better. this paragraph felt a bit big so I was craving a little vertical whitespace.

**Comment by arschles** [comment]
Requested changes made. @duglin @martinmaly @pmorie PTAL again.

@krancour I'd like to leave those title & structural changes to a new PR. Also, would you care to create an issue to describe what the cheat sheet should contain? I'm not following you enough to describe it myself in an issue.

**Comment by martinmaly** [comment]
My `LGTM1` still holds. Thanks for doing this and working through all the feedback!

**Pull request merged**